lib/handleRequest.js

```diff
@@ -125,29 +128,53 @@ function validationCompleted (request, reply, validationErr) {
 function preHandlerCallback (err, request, reply) {
   if (reply.sent) return
 
-  if (err != null) {
-    reply[kReplyIsError] = true
-    reply.send(err)
-    return
+  const context = request[kRouteContext]
+
+  const store = {
+    request,
+    reply,
+    async: false,
+    route: {
+      url: context.config.url,
+      method: context.config.method
+    }
   }
 
-  let result
+  channels.start.runStores(store, () => {
+    try {
+      if (err != null) {
+        reply[kReplyIsError] = true
+        reply.send(err)
+        store.error = err
+        channels.error.publish(store)
+        return
+      }
 
-  try {
-    result = request[kRouteContext].handler(request, reply)
-  } catch (err) {
-    reply[kReplyIsError] = true
-    reply.send(err)
-    return
-  }
+      let result
 
-  if (result !== undefined) {
-    if (result !== null && typeof result.then === 'function') {
-      wrapThenable(result, reply)
-    } else {
-      reply.send(result)
+      try {
+        result = context.handler(request, reply)
+      } catch (err) {
+        store.error = err
+        channels.error.publish(store)
+
+        reply[kReplyIsError] = true
+        reply.send(err)
+        return
+      }
+
+      if (result !== undefined) {
+        if (result !== null && typeof result.then === 'function') {
+          store.async = true
+          wrapThenable(result, reply, store)
+        } else {
+          reply.send(result)
+        }
+      }
+    } finally {
+      channels.end.publish(store)
     }
-  }
+  })
```

Comment:
We were _very_ careful of not creating an added closure in this function. Closures creates overhead and this is _very_ hot. This should removed.

Reply:
`runStores()` accepts arguments to be passed to the wrapped function. So we could use at L143 something like:
```javascript
channels.start.runStores(store, _preHandlerCallback, this, err, request, reply)
```
but looking at the implementation in `dc-polyfill`, it just creates a closure internally: https://github.com/DataDog/dc-polyfill/blob/3a018599e2745ea3373ed151ab525d80f286e669/patch-channel-store-methods.js#L36
So I'm guessing that defeats the point.

Reply:
@mcollina I removed the closure in Fastify but Simon is right in that one still exists within diagnostics channel.

Reply:
@tlhunter Is it even possible to get rid of all closure in a case like this? Node's ALS has `enterWith()` but I'm not sure it's faster than a closure and it's not been kept by the tc39 spec anyway :/

Reply:
@simon-id one thing I could do is check all five channels for listeners. If none are present then run `preHandlerCallbackInner()` without the store argument and without first calling `runStores()`. This prevents the closure in most Fastify use-cases.

Reply:
Pretty sure I fixed this issue